cql3/statements/create_keyspace_statement.cc

```diff
@@ -107,7 +107,16 @@ future<std::tuple<::shared_ptr<cql_transport::event::schema_change>, std::vector
         auto rs = locator::abstract_replication_strategy::create_replication_strategy(
             ksm->strategy_name(),
             locator::replication_strategy_params(ksm->strategy_options(), ksm->initial_tablets()));
-        if (rs->uses_tablets()) {
+        if (_attrs->get_initial_tablets("NetworkTopologyStrategy", feat.tablets).enabled_via_cql
```

Comment:
Why can't each r.s. check for tablets in its validate_options() virtual method?
I know, that it's not going to happen in this era, but when someone adds tablets support to SimpleStrategy, this place will need to look even uglier

Reply:
> That's because of how ks_prop_defs::get_initial_tablets is implemented, and specifically because it skips setting tablets for all strategies other than network, 

True, but there's a fixme there saying that it's not nice either

> and that's because the code later in the processing path is written in such a way that whenever a r.s. has tablets options set, tablets are being created for this r.s.

Can you elaborate on this -- where exactly this happens?
I thought that if ks_prop_defs sets initial-tables for non-NetworkTopology strategy, then r.s.::validate_options() would notice that tablets are requested and throw error about unsupported parameter

Reply:
> I thought that if ks_prop_defs sets initial-tables for non-NetworkTopology strategy, then r.s.::validate_options() would notice that tablets are requested and throw error about unsupported parameter

Only n.t.s. `validate_options` check tablets, other strategies don't do it because they don't have tablets set, because `ks_prop_defs::get_initial_tablets` skips setting them for r.s. other than n.t.s.

>> and that's because the code later in the processing path is written in such a way that whenever a r.s. has tablets options set, tablets are being created for this r.s.
>
> Can you elaborate on this -- where exactly this happens? I thought that if ks_prop_defs sets initial-tables for non-NetworkTopology strategy, then r.s.::validate_options() would notice that tablets are requested and throw error about unsupported parameter

I was a bit wrong, one can create a KS with SimpleStrategy when tablets are turned on (the very essence of this patch is to prevent/warn about it), but later, even if I create a table within such keyspace, system.tablets has no entries for this ks, meaning no tablets are created:
```
cqlsh> CREATE KEYSPACE smaron WITH REPLICATION = {'class': 'SimpleStrategy', 'replication_factor': 3};

Warnings :
SimpleStrategy replication class is not recommended, but was used for keyspace smaron. You may suppress this warning by delisting SimpleStrategy from replication_strategy_warn_list configuration option, or make it into an error by listing this replication strategy on replication_strategy_fail_list.

cqlsh> SELECT * FROM system_schema.scylla_keyspaces  WHERE keyspace_name = 'smaron';

 keyspace_name | initial_tablets | storage_options | storage_type
---------------+-----------------+-----------------+--------------
        smaron |               0 |            null |         null

(1 rows)
cqlsh> CREATE TABLE smaron.my_table (id VARCHAR PRIMARY KEY, field_a VARCHAR, field_b VARCHAR);
cqlsh> SELECT count(*) from system.tablets where keyspace_name = 'smaron' ALLOW FILTERING ;

 count
-------
     0

(1 rows)

Reply:
> Why can't each r.s. check for tablets in its validate_options() virtual method? I know, that it's not going to happen in this era, but when someone adds tablets support to SimpleStrategy, this place will need to look even uglier

SimpleStrategy is already deprecated and we're not going to support tablets with it.
We should even refuse creating new keyspaces with SimpleStrategy.

EverywhereStrategy can be done with tablets.
Though we may want to support same functionality with a "wildcard" per-dc like `{'dc': 'all'}` so it can be implemented as a natural extension to NetworkTopologyStrategy.

Reply:
> SimpleStrategy is already deprecated and we're not going to support tablets with it.
> We should even refuse creating new keyspaces with SimpleStrategy.

Yes, I picked SimpleStrategy just as an example. Read my question as

"I know, that it's not going to happen in this era, but when someone adds tablets support to WheteverStrategy, this place will need to look even uglier"

or as

"I know, that it's not going to happen in this era, but when someone invents SomeOtherStrategy that is better that N.T.Strategy and adds tablets support to it, this place will need to look even uglier"

Reply:
Going back to the original problem, @xemul this comment is heavily connected with [this comment](https://github.com/scylladb/scylladb/pull/20088#discussion_r1746553311). This conversation ends with me saying:
> `ks_prop_defs::get_initial_tablets` skips setting them for r.s. other than n.t.s.

and the linked comment explains why I am recluctant to change it (tl;dr - I tried it, didn't work out).